src/websockethandler.cpp

```diff
@@ -0,0 +1,221 @@
+
+  m_pingTimer = new QTimer(this);
+  connect(m_pingTimer, &QTimer::timeout, this, &WebSocketHandler::ping);
+}
+
+// static
+QString WebSocketHandler::webSocketServerUrl() {
+  QString httpServerUrl;
+  if (Constants::inProduction()) {
+    httpServerUrl = Constants::API_PRODUCTION_URL;
+  } else {
+    httpServerUrl = Constants::getStagingServerAddress();
+  }
+
+  return httpServerUrl.toLower().replace("http", "ws");
+}
+
+// static
+bool WebSocketHandler::isUserAuthenticated() {
+  MozillaVPN* vpn = MozillaVPN::instance();
+  Q_ASSERT(vpn);
+
+  if (vpn->userState() == MozillaVPN::UserAuthenticated) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+void WebSocketHandler::initialize() {
+  logger.debug() << "Initialize";
+
+  MozillaVPN* vpn = MozillaVPN::instance();
+  Q_ASSERT(vpn);
+
+  connect(vpn, &MozillaVPN::userStateChanged, this,
+          &WebSocketHandler::onUserStateChanged);
+}
+
+/**
+ * @brief Opens the websocket connection.
+ *
+ * No-op in case the connection is already open.
+ */
+void WebSocketHandler::open() {
+  if (m_webSocket.state() != QAbstractSocket::UnconnectedState &&
+      m_webSocket.state() != QAbstractSocket::ClosingState) {
+    logger.debug()
+        << "Attempted to open a WebSocket connection, but it's already open.";
+    return;
+  }
+
+  logger.debug() << "Attempting to open WebSocket connection."
+                 << WebSocketHandler::webSocketServerUrl();
+  m_reconnectOnClose = true;
+
+  QNetworkRequest request;
+  request.setRawHeader("Authorization",
+                       SettingsHolder::instance()->token().toLocal8Bit());
+  request.setUrl(QUrl(WebSocketHandler::webSocketServerUrl()));
+  m_webSocket.open(request);
+}
+
+/**
+ * @brief Closes the websocket connection.
+ *
+ * No-op in case the connection is already closed.
+ *
+ * Closing the websocket connection through this function does keeps it closed,
+ * unless the connection is explicitly reopened through
+ * `WebSocketHandler::open`.
+ */
+void WebSocketHandler::close() {
+  if (m_webSocket.state() == QAbstractSocket::UnconnectedState ||
+      m_webSocket.state() == QAbstractSocket::ClosingState) {
+    logger.debug()
+        << "Attempted to close a WebSocket, but it's already closed.";
+    return;
+  }
+
+  logger.debug() << "Closing WebSocket";
+  m_webSocket.close();
+}
+
+/**
+ * @brief Sends a ping to the WebSocket server.
+ *
+ * WebSocket connections may silently be broken. Periodically pinging the server
+ * keeps track of the connection. If verified that the connection has been
+ * broken the connection is properly closed, which triggers new connection
+ * attempts.
+ */
+void WebSocketHandler::ping() {
+  // Never got a pong for the last ping sent, assume connection is broken.
+  if (m_waitingForPong) {
+    logger.debug() << "WebSocket was disconnected silently. Closing.";
+    m_webSocket.close();
+    return;
+  }
+
+  m_waitingForPong = true;
```

Comment:
This approach would not work. When the WebSocket is silently disconnected, the `m_webSocket.state()` will still be `Connected` even though the connection is lost. (I just verified that this is indeed the case, by testing locally. Network disconnection is something that will cause a silent disconnection, for example).

I cannot think of another approach to this issue that does not involve keeping some sort of connection state on the WebSocketHandler. I am 100% open for suggestions though :) Maybe we can use an enum instead of boolean?

Reply:
Hum, ok I re-read your comment and I am wrong! The approach should work indeed because we are closing the connection when there is a timeout. Alright. I'll go change it 🎉